File: exercises/ddd_application_layer/typescript/src/domain/application/ApplicationId.ts
import crypto from 'crypto';

export class ApplicationId {
  private readonly _value: string;

  get value() { return this._value; }

  static generate() {
    return new ApplicationId(crypto.randomUUID());
  }

  constructor(value: string) {
    if (!isValidUUIDv4(value)) {
      throw new Error(`Invalid ApplicationID (must be UUIDv4 format, got ${value})`);
    }
    this._value = value;
  }

  equals(other: ApplicationId) {
    if (!(other instanceof ApplicationId)) {
      return false;
    }
    return this._value === other._value;
  }
}

function isValidUUIDv4(value: string): boolean {
  const uuidV4Regex = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;
  return uuidV4Regex.test(value);
}

File: exercises/ddd_application_layer/java/src/main/java/domain/application/ApplicationId.java
package domain.application;
import java.util.UUID;

public class ApplicationId {
    private UUID value;

    public static ApplicationId generate() {
        return new ApplicationId(UUID.randomUUID());
    }

    public ApplicationId(UUID uuid) {
        this.value = uuid;
    }

    public UUID getValue() {
        return value;
    }

    public boolean equals(ApplicationId other) {
        return this.value.equals(other.value);
    }
}

